drivers/usb/usbxhci/usbxhci.c

```diff
@@ -0,0 +1,572 @@
+     * HC on failure although a delay
+     * could be for many reasons e.g
+     * HC quirks.
+     *
+     * XXX: Should probably handle quirks
+     */
+    KeQuerySystemTime(&EndTime);
+    EndTime.QuadPart += XHCI_POLL_TIME_SET(24);
+    IsPolling = TRUE;
+
+    while (IsPolling)
+    {
+        Status.AsULONG = READ_REGISTER_ULONG(&OperRegisters->UsbStatus.AsULONG);
+        switch (XHCI_PollTimeout(&CurrentTime, &EndTime, Status.HcHalted))
+        {
+            case POLL_STATUS_DONE:
+                IsPolling = FALSE;
+                break;
+            case POLL_STATUS_TIMEOUT:
+                DPRINT1("XHCI_HaltController: Timeout while halting HC\n");
+                return MP_STATUS_ERROR;
+        }
+    }
+
+    DPRINT("XHCI_HaltController: HC is now in a halted state\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+XHCI_ResetController(_In_ PXHCI_HC_OPER_REGS OperRegisters)
+{
+    XHCI_USB_COMMAND Command;
+    XHCI_USB_STATUS Status;
+    LARGE_INTEGER EndTime;
+    LARGE_INTEGER CurrentTime;
+    BOOLEAN IsPolling;
+
+    Command.AsULONG = READ_REGISTER_ULONG(&OperRegisters->UsbCmd.AsULONG);
+    Command.HcReset = 1;
+
+    /* The HC should be halted before a reset */
+    XHCI_HaltController(OperRegisters);
+
+    DPRINT("XHCI_ResetController: RESET\n");
+    WRITE_REGISTER_ULONG(&OperRegisters->UsbCmd.AsULONG,
+                         Command.AsULONG);
+
+    /*
+     * The spec does not specify
+     * how long we must wait before
+     * the HC becomes ready again.
+     * 25ms should suffice.
+     */
+    KeQuerySystemTime(&EndTime);
+    EndTime.QuadPart += XHCI_POLL_TIME_SET(25);
+    IsPolling = TRUE;
+
+    while (IsPolling)
+    {
+        Status.AsULONG = READ_REGISTER_ULONG(&OperRegisters->UsbStatus.AsULONG);
+        switch (XHCI_PollTimeout(&CurrentTime, &EndTime, !Status.ControllerNotReady))
+        {
+            case POLL_STATUS_DONE:
+                IsPolling = FALSE;
+                break;
+            case POLL_STATUS_TIMEOUT:
+                DPRINT1("XHCI_ResetController: Timeout while resetting HC\n");
+                return MP_STATUS_ERROR;
+        }
+    }
+
+    DPRINT("XHCI_ResetController: Reset complete\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+XHCI_InitController(
+    _In_ PXHCI_HC_OPER_REGS OperRegisters,
+    _In_ PUSBPORT_RESOURCES Resources,
+    _In_ PVOID XhciExtension,
+    _In_ PXHCI_HC_CAPABILITY_REGISTERS CapabilityRegisters,
+    _In_ UCHAR CapLength)
+{
+    PXHCI_HC_RESOURCES HcResources;
+    PXHCI_EXTENSION XhciExt;
+    XHCI_HC_STRUCTURAL_PARAMS1 StructParams1;
+    XHCI_HC_CONFIG Config;
+    ULONG DcbaaBasePA;
+    ULONG CommandRingBasePA;
+    ULONG_PTR BaseVA;
+    ULONG_PTR BasePA;
+    UCHAR MaxDeviceSlots;
+    MPSTATUS RetStatus;
+
+    BaseVA = Resources->StartVA;
+    BasePA = Resources->StartPA;
+
+    HcResources = (PXHCI_HC_RESOURCES)BaseVA;
+    DcbaaBasePA = (ULONG)BasePA + FIELD_OFFSET(XHCI_HC_RESOURCES, Dcbaa);
```

Comment:
Physical addresses should always be 64 bit (there's a PHYSICAL_ADDRESS type you could use)